### DISCUSSION TITLE
Board Will Not Register Joystick Input

### DISCUSSION THREAD

**Discussion by Leg-154** [gizmo-platform/gizmo#18] [Q&A]
### Hello! I am reaching out because of some roadblocks I have been hitting when attempting to register joystick inputs with my boards.

Currently, I have been able to get the board to accept programs that allow for program-controlled motor and servo operation, but in attempting to connect the Waveshare with all correct firmware installations, I have not been successful in using the example code basicrobot to register input.

I have a feeling it has to do with firmware issues, because at first I realized I had not flashed the system processor and made sure my UF2 board was paired to the Waveshare with the status LEDs, but some things did go wrong in the firmware installing process.

At some point, I had accidentally flashed the student processor and I attempted a force reset by holding BOOTSEL and plugging it back in (and for some reason it will only let me program the student processor if the rest of the board is unpowered), but nothing still seems to be working.

**Comment by barulicm** [comment]
When you have everything set up and are expecting to see joystick input, what color are the three large system status lights in the corner of the board? Top to bottom (where the programming USB ports are the "top" of the Gizmo), they should be green (network), flashing white (field), and green (battery). We can get into some deeper debugging, but I just want to double check the overall state your system is in first.

A note on programming: holding down the boot select button while plugging in the USB cable doesn't work when the Gizmo is powered on because the Pico is not actually being reset. The Gizmo board is powering it. To put either pico in programming mode while the Gizmo is powered on, you'll need to hold the boot select button while tapping the reset button for that pico. In the enclosure, you'll need a paper clip or other thin poking tool to press the reset button.

**Reply by Leg-154** [reply]
The three status LEDs do match the pattern of green, flashing white, and green. Occasionally, C3 will become yellow and the connection will get a bit fuzzy between the boards, but it usually returns to this same pattern after a minute or two.

Thank you for that second tip! I reset the student Pico with both BOOTSEL and the student reset on the Gizmo board, and re-uploaded the code, but now it seems my joystick won't turn on, even with a powered and connected Waveshare.

**Reply by barulicm** [reply]
Hmmm that behavior of dipping back to yellow on the network status light is unusual. It almost sounds like your driver station might be rebooting periodically. Yellow indicates the Gizmo is connected to the WiFi network but not receiving data from the driver station. However, there's a timeout on the connection status before the Gizmo will admit the network is completely disconnected, so the driver station could completely cycle before the Gizmo status light would have turned red.

When you say your joystick "won't turn on," do you mean the driver station is not powering on or something else?

The driver station's ethernet port has two LEDs (one green and one orange) that should be lit if the driver station is powered on. There's also a green LED on the Raspberry Pi inside the driver station which is visible if you pop the lid off the enclosure.

The joystick itself doesn't really have an "on" state beyond being connected and powered over USB. You can double check by tapping the "Mode" button in the middle, which should toggle the small green light next to the button.

**Reply by Leg-154** [reply]
Apologies for the late reply! The semester has started which made my schedule a lot more difficult to maneuver.

For the joystick neglecting to power on, I was referring to the green light that turns on when pressing "mode".

After plugging in the Waveshare again, i can confirm the LED on the Raspberry Pi does turn on, but the ethernet port LEDs do not seem to do the same. I believe they turned on when I first powered this Waveshare, so I am unsure if there has been a disconnection of some kind.

EDIT: I hooked up the Gizmo and the connection LED does not seem to be having that same odd flickering to yellow, so it may have been something small. The Ethernet LEDs on the Waveshare still remain off, however.

EDIT #2: I ended up replacing this waveshare with a different one, and these lights are now on (alongside normal status LED behavior). I re-uploaded the basicRobot program just to be sure that wasn't the issue, but I am still getting no input.

**Reply by barulicm** [reply]
Ok, if your Gizmo status lights look happy and the joystick + driver station appear to be on, we can next try to determine if the driver station is sending joystick packets to the Gizmo.

We can do this by inspecting the debug logs from the system processor. Connect the system processor to your computer with the programming cable and turn the Gizmo and driver station on. (Note: Do NOT hold the BOOTSEL button. We don't want the processor in programming mode.) With that connected, go to https://www.serialterminal.com/. Click the "Connect" button in the upper left corner and choose the port connected to the Gizmo. It usually shows up as "Pico W" in the device list.

You can also use the serial monitor in the Arduino IDE if you have that on your computer to view the logs.

Once the serial connection is established, you should see logs show up with lines starting with "GIZMO_". We need to check if you see this message: "GIZMO_MQTT_MSG_CONTROL"

If you **do** see GIZMO_MQTT_MSG_CONTROL, then the packets are being sent and it's likely an issue getting that data from the system processor to the student processor.

If you **do not** see GIZMO_MQTT_MSG_CONTROL, then the driver station is not sending joystick packets to the Gizmo. 

A few other questions whose answers might be useful:
* Which version of the Gizmo firmware and driver station image are you using?
* Which programming language are you using for the Basic Robot example? Have you modified the example code in any way?
* Do you have multiple Gizmo systems? Are you seeing this behavior on multiple systems or just one?
* Have you tried running the JoystickDebug example from the Arduino package? It prints the joystick inputs to the serial port. This would confirm that the joystick input is the actual problem, as opposed to an issue with the motors or servos that would be responding to the joystick input.

**Comment by Leg-154** [comment]
I have hooked up the serial monitor to the system processor with the Waveshare turned on as instructed, and this is the response I am getting:
```
GIZMO_NET_WIFI_ATTEMPT
GIZMO_NET_WIFI_START
GIZMO_NET_IP 10.0.0.3
GIZMO_NET_DNS 10.0.0.1
GIZMO_NET_WIFI_RUN
GIZMO_FMS_DISCOVERED_DS 10.0.0.1
GIZMO_MQTT_CONNECT_OK
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
GIZMO_MQTT_MSG_LOCATION
```

With the final message being repeated in an endless loop. Even after rebooting the Waveshare and watching it reconnect, I have not seen any instance of the "GIZMO_MQTT_MSG_CONTROL" packet.

### As for the additional questions, 

- The Gizmo firmware version I am using is gss_R6E_v0.0.4.uf2, and the driver's station firmware I have it flashed with is the drivers-station.zip file on the releases page.
_EDIT: I have now tested the serial monitor with all three available firmware versions R3B and R4B and all yield the same result._
- I am using Arduino, and I had modified the basicRobot program at first just to change the motors to servos, changing all instance of GIZMO_MOTOR_X to GIZMO_SERVO_X, but I have since tested with motors and an unchanged file, and I am having the same issue.
- I have been using the same Gizmo system recently, as before I had an outdated board, and have not yet tested with additional Gizmos.
- I have tested joystickdebug with no code altered, and the serial monitor displays the following in an endless loop, regardless of the buttons pressed:
```
Axis Data: 127 127 127 127 127 127; Button Data: 0 0 0 0 0 0 0 0 0 0 0 0
```

**Reply by barulicm** [reply]
Thanks for the extra info. It looks like your driver station isn't able to read the joystick and send control packets.

Can you check the back of your joystick for the switch labeled X on one side and D on the other? This switch must be in the D position to work with the driver station.

**Reply by Leg-154** [reply]
The controller was in X position, and I went ahead and switched it to D position and began to use JoystickDebug to try and get some input, as well as monitoring the serial output on the system processor, but sadly it still seems to all be sending the same packet with no response. I tried with all three USB ports, as well.

EDIT: if you don't mind me asking, is there a significant difference out of the three firmware versions used to flash the system processor? I haven't found anything that mentions using one version over another.

**Reply by Vaelatern** [reply]
The different board revision for your board. You can find your board revision in the top left of the gizmo board unless it's in a case.

**Reply by Leg-154** [reply]
I have revision H4 v01.00.

**Reply by the-maldridge** [reply]
This is related to gizmo-platform/gizmo#20 which is currently being investigated.

**Reply by the-maldridge** [reply]
Also to answer the question of the three firmware versions, they are all the same software but belong to different generations of hardware which don't all have the same I/O lines dedicated to the same things.  Its important to load the right version on the right board since while they'll all appear to link up and connect, they won't all drive the right signals to the right peripherals.  Notably prior to r6e there was no ethernet, and in the version 1.0 firmware that hasn't been pushed yet, ethernet can operate in interrupt mode instead of polling mode.
